Adapted `apps.md` from ABCI directory (#8506)
tendermint/tendermint: spec/abci++/abci++_app_requirements_002_draft.md

  but may also depend on other values or operations.
* *w<sup>r</sup><sub>p</sub> = w<sup>r</sup><sub>q</sub> &#8655;
  e<sup>r</sup><sub>p</sub> = e<sup>r</sup><sub>q</sub>*

## Managing the Application state and related topics

### Connection State

Tendermint maintains four concurrent ABCI++ connections, namely
[Consensus Connection](#consensus-connection),
[Mempool Connection](#mempool-connection),
[Info/Query Connection](#infoquery-connection), and
[Snapshot Connection](#snapshot-connection).
It is common for an application to maintain a distinct copy of
the state for each connection, which are synchronized upon `Commit` calls.

#### Concurrency

> In principle, each of the four ABCI++ connections operates concurrently with one

Author: I took this from the original text in ABCI spec. Has this changed recently?

another. This means applications need to ensure access to state is
thread safe. Up to v0.35.x, both the
[default in-process ABCI client](https://github.com/tendermint/tendermint/blob/v0.35.x/abci/client/local_client.go#L18)
and the
[default Go ABCI server](https://github.com/tendermint/tendermint/blob/v0.35.x/abci/server/socket_server.go#L32)

> used global locks to guard the handling of events across all connections, so they were not

Reviewer: ```suggestion
used a global lock to guard the handling of events across all connections, so they were not
```

concurrent at all. This meant if your app was written in Go, and compiled in-process with Tendermint

> using the default `NewLocalClient`, or run out-of-process using the default `SocketServer`,

Author: Understood. This was adapted from the original text

> ABCI messages from all connections were linearizable (received one at a

Author: I have the same though, but decided to leave it. Since you're also not sure about the term, let my try to adapt the text

time).

> This is no longer the case starting from v0.36.1: the global locks have been removed and it is

Reviewer: ```suggestion
This is no longer the case starting from v0.36.0: the global locks have been removed and it is
```

> up to the Application to ensure that handling ABCI++ connections is thread safe.

Reviewer: Just to be more specific about what "thread safe" means in this context:
```suggestion
up to the Application to synchronize access to its state when handling
ABCI++ methods on all connections.
```

#### FinalizeBlock

When the Consensus algorithm decides on a block locally, Tendermint uses `FinalizeBlock` to send the

> current block's hash, height, transactions, etc., which the Application uses to transition its state.

Author: Agreed, will change to "decided".
We don't disclose the header any longer, but relevant fields of it. Also, transactions, etc.

> current block's hash, height, transactions, etc., which the Application uses to transition its state.

Author: removed

> The Application should remember the latest height from which it

Reviewer: Not sure if this is a convenience ("should") or a requirement ("must").
```suggestion
The Application must remember the latest height from which it
```

Author: Good point!

has run a successful `Commit` so that it can tell Tendermint where to

> pick up from when it restarts. See information on the Handshake, below.

Author: Fair enough

> pick up from when it restarts. See information on the Handshake, below.

Author: Added link

#### Commit

> The Application state should be persisted during `Commit`.

Author: Good point

> Before invoking `Commit`, Tendermint locks the mempool and flushes the mempool connection. This ensures that
> no new messages

Reviewer: ```suggestion
Before invoking `Commit`, Tendermint blocks the mempool and flushes the mempool connection. This ensures that
no new messages
```

Author: I think "locks" is more precise here

> will be received on the mempool connection during this processing step, providing a good opportunity to safely update all four

Reviewer 2: there is only one connection, and the mempool connection is not separate from the other connections.

> connection states to the latest committed state at once.

Reviewer 2: "at once" indicates timliness, but I think "atomically" is actually closer to what you mean or want.

Author: changed to "at the same time"

> When `Commit` completes, Tendermint unlocks the mempool.

Reviewer 2: I would say "returns" rather than "completes"

Author: Yep

> WARNING: if the ABCI app logic processing the `Commit` message sends a

Reviewer: We could probably detect this and crash for it (e.g., check in the ABCIQuery handler whether a commit is pending). That would help ensure applications don't try to do this.

> `/broadcast_tx_sync` or `/broadcast_tx_commit` and waits for the response

Reviewer 2: I would remove _comit entirely from the documentation as it's deprecated.

Author: OK, did a 's/_commit//g' in this file

before proceeding, it will deadlock. Executing those `broadcast_tx` calls
involves acquiring a lock that is held during the `Commit` call, so it's not

> possible. If you make the call to the `broadcast_tx` endpoints concurrently,

Author: Let me see what I can do

that's not a problem, it just can't be part of the sequential logic of the
`Commit` function.

#### Candidate States

> Tendermint calls `PrepareProposal` when it is about to send a proposed block to the network.

Reviewer 2: `is about to` -> `before`

Author: "before" is ambiguous here. It could be 1/2 hour earlier. I'll keep "is about to"

> Likewise, Tendermint calls `ProcessProposal` upon reception of a proposed block from the

Reviewer 2: ```suggestion
Tendermint calls `ProcessProposal` after receiving a proposed block from the
```

Author: "upon" denotes cause-effect. "after" is just temporal ordering

> network. In both cases, the proposed block's data (last block's commit info, transactions, etc.)

Reviewer 2: It would be more clear (and in general) to omit the parenthetical statements.

is disclosed to the Application, in the same conditions as is done in `FinalizeBlock`.

The Application may decide to *immediately* execute the given block (i.e., upon `PrepareProposal`
or `ProcessProposal`). There are two main reasons why the Application may want to do this:

> * In order to be sure that the block does not contain *any* invalid transaction, there is usually

Reviewer 2: "usually? is unclear.

Author: "usually", because there may be cases of (simple) applications that don't need to fully execute the transactions to detect the invalid ones.
Anyway, I reworded

> * In order to be sure that the block does not contain *any* invalid transaction, there is usually
>   no way other than fully executing the transactions in the block as though it was the *decided*
>   block.

Reviewer: I'm not sure this is "usual". It seems very much dependent upon what the application defines its transactions to be.
```suggestion
* In some cases, an application cannot tell whether all the transactions in a block
  are valid without executing them as though it was the decided block.
```

> * Upon reception of the decided block via `FinalizeBlock`, if that same block was executed
>   upon `PrepareProposal` or `ProcessProposal` and the resulting state was kept in memory, the
>   Application can simply apply that state (faster) to the main state, rather than reexecuting
>   the decided block (slower).

Reviewer: Suggestion to front-load the key point:
```suggestion
* By executing the block ahead-of-time (during `PrepareProposal` or `ProcessProposal`),
  the application can cache the new state and apply immediately if and when that block 
  is finalized. This allows the node to proceed to the next block more quickly.
```

> `PrepareProposal`/`ProcessProposal` can be called many times for a given height. Moreover,

Reviewer 2: ```suggestion
`PrepareProposal` and `ProcessProposal` are safe to call more than once for a given height. Moreover,
```

Author: I'll keep the original wording. This is aimed at App devs, not Tendermint devs, so the fact that it is "safe to call" `PrepareProposal` or `ProcessProposal` several times only concerns Tendermint.
The App just needs to know it can happen.

> `PrepareProposal`/`ProcessProposal` can be called many times for a given height. Moreover,
> it is not possible to accurately predict which of the blocks proposed in a height will be decided, being delivered to the application
> in that height's `FinalizeBlock`.
> Therefore, the state resulting from executing a proposed block, denoted a *candidate state*, should
> be kept in memory as a possible final state for that height. When `FinalizeBlock` is called, the Application should
> check if the decided block corresponds to one of its candidate states; if so, it will apply it as
> its `ExecuteTxState` (see [Consensus Connection](#consensus-connection) below),
> which will be persisted during the upcoming `Commit` call.

Reviewer: Suggestion to front-load the key point:
```suggestion
Blocks executed during the `PrepareProposal` or `ProcessProposal` stages must be
treated as *candidate states* and must not affect the user-visible state of the
application. The `PrepareProposal` and `ProcessProposal` methods can be called multiple
times for a given height (once per round), and it is not possible to know in advance
which of the blocks proposed in a height will be decided.

Only when `FinalizeBlock` is called may the Application apply a candidate state computed
for that block to its `ExecuteTxState` (see [Consensus Connection](#consensus-connection) below),
to be persisted during the following `Commit` call.
```

Also, "in memory" is an implementation detail and probably not necessary to specify. (If an application wants to cache candidate states on disk, that would be fine too).

> In bad runs (e.g., under network instability), Consensus might take many rounds.

Reviewer 2: I would opt for something that's more clear than "bad runs". 

I'm also worried about the the phrasing of "Consensus" as an actor that is part of the specification (rather than just saying "Tendermint"? I'm not sure that people need to think about our internal state machine.

Author: OK, reworded

In this case, several proposed blocks will be disclosed to the Application for a given height.

> Indeed, by the nature of Consensus, the number of proposed blocks received by the Application for

Reviewer 2: ```suggestion
The number of proposed blocks received by the Application for
```

Author: But I want give a hint on _why_ the rounds can't be bound...
I'll remove "Indeed", though

a particular height cannot be bound, so Application developers must act with care and use mechanisms
to bound memory usage. As a general rule, the Application should be ready to discard candidate states
before `FinalizeBlock`, even if one of them might end up corresponding to the

> decided block and thus have to be reexecuted upon `FinalizeBlock`.

Reviewer: I think this paragraph mostly restates what the previously one said, and could be reduced, combined, or eliminated.

### States and ABCI++ Connections

#### Consensus Connection

> The Consensus Connection should maintain an `ExecuteTxState` - the working state

Reviewer: The typography `ExecuteTxState` seems to imply this is a data structure defined by Tendermint, but I think this is an application implementation detail. I think the key point is something like:

> The application must track which candidate execution state will be reported as the latest committed state during the `Commit` operation on the consensus connection. This value is updated by the call to `FinalizeBlock`, which identifies the decided block. Block executions computed during `PrepareProposal` or `ProcessProposal` **must not** update this state.

or words to that effect. Or maybe just call out specifically that `ExecuteTxState` is not a specific data structure that we're identifying, it's a notional thing we're asserting the application has to keep track of in whatever way makes sense.

Author: I'll keep the terms ("ExecuteTxState", "CheckTxState", etc) because they're referenced at several places in text later on. So I don't want to get into coming up with a complete new way of explaining all this.
Having said that, I agree with you that the typography used is confusing in that it matches the one used for _actual_ data structures or method names. I'll change it to italic. I think that'll eliminate the confusion.

for block execution. It should be updated by the call to `FinalizeBlock`
during block execution and committed to disk as the "latest
committed state" during `Commit`. Execution of a proposed block (via `PrepareProposal`/`ProcessProposal`) **must not** update the `ExecuteTxState`, but rather be kept as a separate candidate state until `FinalizeBlock`

> confirms which of the candidate states (if any) can be used to update `ExecuteTxState`.

Reviewer 2: I'd say something like "calls to FinalizeBlock should update the working state during block execution. 

I'm not sure that we should specify the latest comitted state as part of the specification? isn't this just an expedience?

#### Mempool Connection

The mempool Connection should maintain a `CheckTxState`
to sequentially process pending transactions in the mempool that have

> not yet been committed. It should be initialized to the latest committed state

Author: reset

at the end of every `Commit`.

During the execution of a Consensus instance, the `CheckTxState` may be updated concurrently with the
`ExecuteTxState`, as messages may be sent concurrently on the Consensus and Mempool connections.
At the end of the Consensus instance, as described above, Tendermint will lock the mempool and flush
the mempool connection before calling `Commit`, ensuring that all existing `CheckTx` calls are
responded to and no new ones can begin.

> After `Commit`, while still holding the mempool lock, `CheckTx` is run again on all transactions

Author: > This is not mandatory though, as it can be disabled by a config flag.
Yes but do we really want to include that detail here?

> After the commit call returns, is that what you mean?
Yes. Will reword

that remain in the node's local mempool after filtering those included in the block.

> An additional `Type` parameter is made available to the `CheckTx` function that

Author: Yeah should have reworded this (probably comes from the time it was introduced).
I will "flatten" the expression

indicates whether an incoming transaction is new (`CheckTxType_New`), or a
recheck (`CheckTxType_Recheck`).

Finally, after re-checking transactions in the mempool, Tendermint will unlock

> the mempool connection. New transactions are once again able to be processed through `CheckTx`.

Reviewer 2: there isn't a specific mempool connection, and this is confusing as a result.

Author: Again, this is taken from the original ABCI spec. Did something change architecturally in the meantime?

Note that `CheckTx` is just a weak filter to keep invalid transactions out of the blockchain.
Since the transaction cannot be guaranteed to be checked against the exact same state as it
will be executed as part of a (potential) decided block, `CheckTx` shouldn't check *everything*
that affects the transaction's validity, in particular those checks whose validity may depend on
transaction ordering. `CheckTx` is weak because a Byzantine node need not care about `CheckTx`;
it can propose a block full of invalid transactions if it wants. The mechanism ABCI++ has
in place for dealing with such behavior is `ProcessProposal`.

##### Replay Protection

> To prevent old transactions from being replayed, `CheckTx` must implement

Reviewer 3: This was written in a pretty clear way somewhere in the docs, but I couldn't find the page...

> To prevent old transactions from being replayed, `CheckTx` must implement

Author: OK, I'll include something along those lines.

replay protection.

It is possible for old transactions to be sent to the application. So

> it is important `CheckTx` implements some logic to handle them.

Reviewer 2: I don't know that this is necessarily true.

Author: This is taken from the original text, but it seems reasonable to me...

Reviewer 4: I think the current TM data store implementation is difficult to figure out if the transaction is an old transaction or not unless tx indexer is enabled and the mempool is able to query it (by txhash?). The App should implement logic to be able to reject the transaction so let it cannot be stored in the mempool.

Author: Yes, as we are talking about `CheckTx` here, we _imply_ that it is in principle up to the Application to implement replay protection.
I'll clarify this in the text

Author: Revamped the whole section

#### Info/Query Connection

> The Info (or Query) Connection should maintain a `QueryState` for answering queries from the user,

Reviewer 3: I would list the two uses of this connection: 1) servicing/answering user queries (although this is a pretty general, not specific definition), and 2) initializing the application (when exactly?).

The use of these queries is not really described below...

Author: The `Query` mechanism (your point 1) is extensively explained below in its own section, so I'll add a forward reference.
I'll add something for 2)

and for initialization when Tendermint first starts up (both described further
below).

> `QueryState` always contain a copy of of the latest `ExecuteTxState` at the end of every `Commit`,

Reviewer 3: Should always? Maybe be a copy/snapshot of the `ExecuteTxState` after the last commit?

Author: Yep, "should" was missing.
Let me reword it closer to what you wrote

after the full block has been processed and the state committed to disk.

> Otherwise it should never be modified.

Reviewer 3: This is not clear. Should it be a read-only copy of the latest committed state?

> Tendermint Core currently uses the Query connection to filter peers upon

Reviewer 3: This is an example of user queries/commands, correct?

Author: Yeah, for some reason, it was included in this section and I didn't dare to delete it.
I guess it's important because it's a "special" example that affects p2p, and not an external user as in the normal case.

Reviewer 2: I'd put it in it's own section. It doesn't really make sense here.

Author: OK, I'll delete it, then

connecting, according to IP address or node ID. For instance,
returning non-OK ABCI response to either of the following queries will
cause Tendermint to not connect to the corresponding peer:

* `p2p/filter/addr/<ip addr>`, where `<ip addr>` is an IP address.
* `p2p/filter/id/<id>`, where `<is>` is the hex-encoded node ID (the hash of
  the node's p2p pubkey).

> Note: these query formats are subject to change!

Reviewer 2: I'd omit this line.

Author: It's now deleted 🙂

#### Snapshot Connection

> The Snapshot Connection is optional, and is only used to serve state sync snapshots for other nodes

Reviewer 3: The connection is not optional, but managing/storing/maintaining snapshots is.

Author: Good point

and/or restore state sync snapshots to a local node being bootstrapped.

For more information, see [the state sync section](#state-sync) below.

### Transaction Results

The Application is expected to return a list of
[`ExecTxResult`](./abci%2B%2B_methods_002_draft.md#exectxresult) in
[`ResponseFinalizeBlock`](./abci%2B%2B_methods_002_draft.md#finalizeblock). The list of transaction
results must respect the same order as the list of transactions delivered via
[`RequestFinalizeBlock`](./abci%2B%2B_methods_002_draft.md#finalizeblock).
This section discusses the fields inside this structure, along with the fields in
[`ResponseCheckTx`](./abci%2B%2B_methods_002_draft.md#checktx),
whose semantics are similar.

The `Info` and `Log` fields are

> non-deterministic values for debugging/convenience purposes that are otherwise ignored.

Reviewer 3: What you mean by "otherwise ignored"?

Author: ... "by Tendermint"

#### Gas

Ethereum introduced the notion of *gas* as an abstract representation of the

> cost of resources used by nodes when processing transactions. Every operation in the

Reviewer 3: ```suggestion
cost of the resources consumed by nodes when processing transactions. Every operation in the
```

> Ethereum Virtual Machine uses some amount of gas, and gas can be accepted at a market-variable price.

Reviewer 3: Accepted? Do you mean "charged"?

> Users propose a maximum amount of gas for their transaction; if the transaction uses less, they get

Reviewer 3: What happens if takes/uses/consumes more?

Author: I've explained it in the re-worded text to your previous comment. Please check

the difference credited back. Tendermint adopts a similar abstraction,
though uses it only optionally and weakly, allowing applications to define
their own sense of the cost of execution.

In Tendermint, the [ConsensusParams.Block.MaxGas](../proto/types/params.proto) limits the amount of gas

> that can be used in a block.

Reviewer 3: More specifically, for the transactions included in a block? Does it also count for something else in a block?

Author: I don't think so. I'll clarify

> The default value is `-1`, meaning no limit, or that the concept of gas is

Reviewer 3: No limit means that gas consumption is not bound, not considered, or both? Meaningless is not clear for me.

Author: "not bound"... the App could still be returning gas for transactions in `CheckTx`.
Will reword

meaningless.

> Responses contain a `GasWanted` and `GasUsed` field. The former is the maximum

Reviewer 3: Execution, validation, or both?

Author: both. Will clarify

amount of gas the sender of a tx is willing to use, and the latter is how much it actually
used. Applications should enforce that `GasUsed <= GasWanted` &mdash; ie. transaction execution

> should halt before it can use more resources than it requested.

Reviewer 3: What does it mean in practice? Transactions can be aborted? What happens with the state they touch?

When `MaxGas > -1`, Tendermint enforces the following rules:

* `GasWanted <= MaxGas` for every transaction in the mempool
* `(sum of GasWanted in a block) <= MaxGas` when proposing a block

If `MaxGas == -1`, no rules about gas are enforced.

In v0.35.x and earlier versions, Tendermint does not enforce anything about Gas in the consensus,
only the mempool.
This means it does not guarantee that committed blocks satisfy these rules.
It is the application's responsibility to return non-zero response codes when gas limits are exceeded.
Since the introduction of `PrepareProposal` and `ProcessProposal` in v.0.36.x, it is now possible
to enforce that all blocks proposed (and voted for) in consensus &mdash; and thus all
blocks decided &mdash; respect the `MaxGas` limits described above.

> The `GasUsed` field is ignored completely by Tendermint. That said, applications should enforce:

Reviewer 3: ```suggestion
The `GasUsed` field is ignored by Tendermint. That said, applications should enforce:
```

* `GasUsed <= GasWanted` for any given transaction
* `(sum of GasUsed in a block) <= MaxGas` for every block

#### Specifics of `ResponseCheckTx`

> If `Code != 0`, it will be rejected from the mempool and hence

Reviewer 3: It will not be added to the mempool, I think. The sender of the transaction will receive an error if the tx was submitted using sync or commit methods.

Author: I prefer the current wording because it could be a "recheck", in that case the tx _was_ already in the mempool.
It could also be coming from p2p.

not broadcasted to other peers and not included in a proposal block.

> `Data` contains the result of the `CheckTx` transaction execution, if any. It is not required to be

Reviewer 3: I would say that is not expected to be deterministic, since checks are made against potentially distinct states.

Author: will change "required" by "expected"

deterministic since, given a transaction, nodes' Applications
might have a different `CheckTxState` when they receive it and check their validity via `CheckTx`.
It is

> semantically meaningless to Tendermint.

Reviewer 3: Again, what do you mean by that? Tendermint ignores, discards, its operation is not affected?

Author: OK, will reword

> `Events` include any events for the execution, though since the transaction has not

Reviewer 3: Not clear.

Author: Yeah... I'm wondering if we shouldn't just drop these parameters. They seem to be there for "symmetry" with `DeliverTx`. But `DeliverTx` is gone.
@cmwaters, what do you think?

Reviewer: Yeah, I think the two methods used to share the same parameter type. But as you say, we don't need this field in `CheckTx` (and never really did, since even in legacy ABCI we did not index ABCI metadata until after delivery).

Reviewer 5: Yeah I agree and was doing some work yesterday to see what fields actually made sense. From my view I think we should drop `Info`, `Events`, `MempoolError` and eventually `Logs` and `GasUsed` (`GasWanted` is the only part the mempool cares about for gas calculations)

Author: @cmwaters did you open an issue when looking into this? (or maybe there is one already?)

been committed yet, they are effectively ignored by Tendermint.

> In v0.35.x and later, there is a `Priority` field in `ResponseCheckTx` that can be

Reviewer 3: ```suggestion
From v0.35.x, there is a `Priority` field in `ResponseCheckTx` that can be
```

used to explicitly prioritize transactions in the mempool for inclusion in a block
proposal.

#### Specifics of `ExecTxResult`

`FinalizeBlock` is the workhorse of the blockchain. Tendermint delivers the decided block,
including the list of all its transactions synchronously to the Application.

> The block delivered (and thus the transaction order) is the same at all correct nodes as guaranteed

Reviewer 3: The same block, this means the same transactions in the same total order, is delivered to all nodes. This is the basic assumption of SMR.

Author: Yes, same transactions. But Tendermint agrees on the block, the transactions' total order (needed for SMR) is a side effect of this.
Other things like the validator set is also agreed, and the total order in changes to the validator set is also necessary for SMR in a dynamic model (i.e., with changing membership).

by the Agreement property of Tendermint consensus.

> In same block execution mode `ExecTxResult` is returned by `PrepareProposal` so that its data can

Reviewer 3: ```suggestion
In same block execution mode, `ExecTxResult` is returned by `PrepareProposal` so that its data can
```

Reviewer 3: Not clear whether this is mandatory or advised.

> be used to build the block to be proposed in the current height.

Author: I have added a few sentences

The `Data` field in `ExecTxResult` contains an array of bytes with the transaction result.

> It must be strictly deterministic (all nodes should return the same value), but can contain arbitrary

Reviewer 3: ```suggestion
It must be deterministic (i.e., the same value must be returned in all nodes), but it can contain arbitrary
```

Reviewer 3: What you mean by arbitrary here? I think the point here is: it can return "arbitrary" data, in the sense that anything the application considers relevant, but at the same time deterministic.

data. Likewise, the value of `Code` must be deterministic.
Both the `Code` and `Data` are included in a structure that is hashed into the

> `LastResultsHash` of the block header in the next height (next block execution mode), or the

Reviewer 3: The two behaviors here should be highlighted, as this is a very important information. I would move the `Code != 0` discussion below to this paragraph, and discuss where this hash pertains to a new paragraph.

header of the block to propose in the current height (same block execution mode, `ExecTxResult` as
part of `PrepareProposal`).

If `ExecTxResult` contains `Code != 0`, the transaction will be considered invalid,

> though it is still included in the block.

Reviewer 3: So, what does it means invalid here? Is this returned to the sender?

Author: Reworded to "marked invalid"

> though it is still included in the block.

Reviewer: That's an interesting question. 

In favour of indexing is that, if you don't, then a transaction that exists in the block will not be discoverable. Whether it's valid or not, it will appear not to have been included.

In favour of not indexing is that it had no effect on the resulting state (presumably), so from the caller's perspective it's _as if_ the transaction hadn't been included.

I cautiously suspect we _should_ index, but I agree we should be explicit either way.

> though it is still included in the block.

Author: My understanding is that a transaction returning a code !=0 shouldn't have been included in the first place. So I'd say it should not be indexed.
With ABCI++ this kind of transaction can be prevented from ever making it to a block. So in this case, the sender won't find it indexed either.
I'll add a sentence

> `Events` include any events for the execution, which Tendermint will use to index

Reviewer 3: Is this specific for `ExecTxResult`?

Author: No, it's in other places (`checkTx` and `FinalizeBlock`), but here we are explaining transaction results, so we should focus in that.

the transaction by. This allows transactions to be queried according to what
events took place during their execution.

### Updating the Validator Set

> The application may set the validator set during

Reviewer 3: The application may configured the validator set in the 3 calls. Details are provided in the following.

Author: I prefer the current wording

[`InitChain`](./abci%2B%2B_methods_002_draft.md#initchain), and may update it during
[`FinalizeBlock`](./abci%2B%2B_methods_002_draft.md#finalizeblock)
(next block execution mode) or
[`PrepareProposal`](./abci%2B%2B_methods_002_draft.md#prepareproposal)/[`ProcessProposal`](./abci%2B%2B_methods_002_draft.md#processproposal)
(same block execution mode). In all cases, a structure of type
[`ValidatorUpdate`](./abci%2B%2B_methods_002_draft.md#validatorupdate) is returned.

> The `InitChain` method can return a list of validators.

Reviewer 3: Suggestion: used to initialize the application <put when here>.

Author: Good idea, included

> If the list is empty, Tendermint will use the validators loaded in the genesis

Reviewer 3: ```suggestion
If the list is empty, Tendermint will use the validators loaded from the genesis
```

Reviewer 3: Question: is the genesis validator set provided in this call? If so, mention it.

file.

> If the list returned by `InitChain` is not empty, Tendermint will use its contents as the validator set.

Reviewer 3: Question: is it a list of validators or validator updates? Is the list *adopted* by Tendermint, or are the operations (updates) performed/implemented? What happens if this list is not well formed?

Author: That's a good question. I'll answer it here, but think getting into the details of this here would be overkill (it should be explained in the "methods" section (for `InitChain`).

The answer: you're right in that it's updates. In normal execution that update contains (a) new validators and their power (the validator MUST NOT exist in the set before the update), (b) existing validators with modified power (power is overwritten on top of the previous one), (c) validators with power set to 0 (those validators MUST exist in the set before update; they are removed from the set).

In this case, we use the same semantics, with the empty set as the set before the update, and the update only containing (a) updates.

So when you look at the update struct (with only (a) updates) it's pretty similar to the set it updates to.

Author: Actually, let me add something like this in the "Methods" section in this PR, even if it's not part of ABCI++ (strictly speaking)

This way the application can set the initial validator set for the
blockchain.

> Note that the maximum total power of the validator set is bounded by

Reviewer 3: Suggestion: use a different format here, start with a "Note" or "Important".

Reviewer 3: Suggestion: re-organize the description so that users know that voting power are "arbitrary", but that the maximum power for a set is bounded. I think they are unlikely to violate this requirement though.

Author: OK, let me see what I can do

`MaxTotalVotingPower = MaxInt64 / 8`. Applications are responsible for ensuring
they do not make changes to the validator set that cause it to exceed this
limit.

> Additionally, applications must ensure that a single set of updates does not contain any duplicates &mdash;

Reviewer 3: ```suggestion
Applications must ensure that a single set of validator updates does not contain duplicates, namely
```

a given public key can only appear once within a given update. If an update includes
duplicates, the block execution will fail irrecoverably.

> Structure `ValidatorUpdate` contains a public key to reference the validator:

Reviewer 3: Suggestion: a validator is defined/referenced/individuated by its public key. Plus, collapse this paragraph with the next.

Author: OK

> The public key currently supports three types:

Reviewer 3: Suggestion: three types of .... are supported (currently is redundant)

Author: I'll leave "currently" nevertheless. If someone reads this in the future and it is outdated, it won't be as "wrong".

Author: "currently" also gives a sense that the list is, by no means, trying to be complete

* `ed25519`
* `secp256k1`
* `sr25519`

> Structure `ValidatorUpdate` also contains an `ìnt64` with the validator's new power.

Reviewer 3: Suggestion: the public key defines the validator, the int field defines its voting power.

Author: I re-worded otherwise

It must abide by the
following rules:

* power must be non-negative

> * if power is 0, the validator must already exist, and will be removed from the

Reviewer 3: Suggestion: If it is set to 0, the validator will be removed. The validator must exist etc. etc.

Author: Actually I tried this, but then "it must exist" becomes ambiguous, since the set is changing, and I need to specify "in the validator set prior to the update"... which is cumbersome

  validator set

> * if power is greater than 0:

Reviewer 3: Suggestion: don't use "exist". If the validator is part, is present on the set etc.

Author: good point... I'll use "is in" as in A &isin; B

    * if the validator does not already exist, it will be added to the validator
      set with the given power
    * if the validator does already exist, its power will be adjusted to the given power

> * the total power of the new validator set must not exceed `MaxTotalVotingPower`

Reviewer 3: Suggestion: this was already mentioned, I would move the paragraph that comments this to after this bullet list.

Author: I see. Let's see if I can do that

Author: In the end, I just deleted the previous paragraph, and integrated the few non-repeated parts over here. Now, there's no repetition

> Note the updates returned in block `H` will only take effect at block `H+2`.

Reviewer 3: Make it clear: they are returned at the end of the processing phase of the block at height X. They will take effect (the validator set?) from block X+Y. Add a link to where this is detailed.

### Consensus Parameters

> `ConsensusParams` enforce certain limits in the blockchain, like the maximum size

Reviewer 3: Suggestion: are global parameters, apply to all validators in a blockchain, control the operation of the Consensus protocol.

Author: (y)

If `storeBlockHeight > stateBlockHeight+1`, panic

Now, the meat:

If `storeBlockHeight == stateBlockHeight && appBlockHeight < storeBlockHeight`,
replay all blocks in full from `appBlockHeight` to `storeBlockHeight`.
This happens if we completed processing the block, but the app forgot its height.

If `storeBlockHeight == stateBlockHeight && appBlockHeight == storeBlockHeight`, we're done.
This happens if we crashed at an opportune spot.

If `storeBlockHeight == stateBlockHeight+1`
This happens if we started processing the block but didn't finish.

If `appBlockHeight < stateBlockHeight`
    replay all blocks in full from `appBlockHeight` to `storeBlockHeight-1`,
    and replay the block at `storeBlockHeight` using the WAL.
This happens if the app forgot the last block it committed.

If `appBlockHeight == stateBlockHeight`,
    replay the last block (storeBlockHeight) in full.
This happens if we crashed before the app finished Commit

If `appBlockHeight == storeBlockHeight`
    update the state using the saved ABCI responses but dont run the block against the real app.
This happens if we crashed after the app finished Commit but before Tendermint saved the state.

> ### State Sync

Reviewer 5: This is great but I feel like it should be it's own page in a different directory in the spec: See https://github.com/tendermint/tendermint/issues/8219

Author: Yes, that's a good idea. Just added a comment in #8219

verification during the restore to detect failures early. This might e.g. include incremental
verification of each chunk against the app hash (using bundled Merkle proofs), checksums to
protect against data corruption by the disk or network, and so on. However, it is important to
note that the only trusted information available is the app hash, and all other snapshot metadata
can be spoofed by adversaries.

Apps may also want to consider state sync denial-of-service vectors, where adversaries provide
invalid or harmful snapshots to prevent nodes from joining the network. The application can
counteract this by asking Tendermint to ban peers. As a last resort, node operators can use
P2P configuration options to whitelist a set of trusted peers that can provide valid snapshots.

##### Transition to Consensus

Once the snapshots have all been restored, Tendermint gathers additional information necessary for
bootstrapping the node (e.g. chain ID, consensus parameters, validator sets, and block headers)
from the genesis file and light client RPC servers. It also fetches and records the `AppVersion`

> from the ABCI application.

Reviewer 5: Statesync calls `Info` to verify two things:
- That the app hash from the collective chunks it has delivered to the app is the same as the apphash recorded by the network (i.e. from the block at the next height)
- That the version that the application sends matches the version of the network

Author: OK, let me see how I can include that info

Author: Done. Please check the new text

Ethereum Virtual Machine uses some amount of gas.
Gas has a market-variable price based on which miners can accept or reject to execute a
particular operation.

Users propose a maximum amount of gas for their transaction; if the transaction uses less, they get
the difference credited back. Tendermint adopts a similar abstraction,
though uses it only optionally and weakly, allowing applications to define
their own sense of the cost of execution.

In Tendermint, the [ConsensusParams.Block.MaxGas](#consensus-parameters) limits the amount of
total gas that can be used by all transactions in a block.
The default value is `-1`, which means the block gas limit is not enforced, or that the concept of
gas is meaningless.

Responses contain a `GasWanted` and `GasUsed` field. The former is the maximum
amount of gas the sender of a tx is willing to use, and the latter is how much it actually
used. Applications should enforce that `GasUsed <= GasWanted` &mdash; i.e. transaction execution
or validation should fail before it can use more resources than it requested.

When `MaxGas > -1`, Tendermint enforces the following rules:

* `GasWanted <= MaxGas` for every transaction in the mempool

> * `(sum of GasWanted in a block) <= MaxGas` when proposing a block

Reviewer 4: If we use this rule to measure how full the block is, how to prevent someone submit transactions with high Gaswanted? So the chain can have very low tx throughput.

Author: This question is only relevant to (successful) blockchains that hit the block limits (almost) every time they reap transactions from the mempool at proposal time; IOW, chains where there is a continued backlog of transactions in the mempool.
The best-known way to deal with contention in this case is through fees, so I imagine that those cases will need the Application to somehow make the gas proportional to the fees paid for a transaction. Therefore, if an attacker sends transactions with a big `GasWanted` value (in order to cause blocks to have few transactions, and thus worsen the mempool backlog), they will have to PAY for that. So the high cost probably defeats the purpose.

Reviewer 4: I see, so the App will be required to implement some transaction fee charge mechanism to prevent this case.

Author: That is my understanding. Bear in mind this hasn't changed with ABCI++. This is the original design, so I think it is now battle-tested.

If `MaxGas == -1`, no rules about gas are enforced.

In v0.35.x and earlier versions, Tendermint does not enforce anything about Gas in the consensus,
only the mempool.
This means it does not guarantee that committed blocks satisfy these rules.
It is the application's responsibility to return non-zero response codes when gas limits are exceeded.
Since the introduction of `PrepareProposal` and `ProcessProposal` in v.0.36.x, it is now possible
to enforce that all blocks proposed (and voted for) in consensus &mdash; and thus all
blocks decided &mdash; respect the `MaxGas` limits described above.

The `GasUsed` field is ignored by Tendermint. That said, applications should enforce:

* `GasUsed <= GasWanted` for any given transaction

> * `(sum of GasUsed in a block) <= MaxGas` for every block

Reviewer 4: Is that possible to use this rule for proposing a block?

Author: Yes, it follows from:
* Tendermint enforcing `(sum of GasWanted in a block) <= MaxGas` in proposed blocks
* Applications enforcing `GasUsed <= GasWanted` (so the Application will stop and fail the tx the moment its execution's `GasUsed` hits `GasWanted`.

I'll explain this in the text

When Tendermint connects to a peer, it sends two queries to the ABCI application
using the following paths, with no additional data:

* `/p2p/filter/addr/<IP:PORT>`, where `<IP:PORT>` denote the IP address and
  the port of the connection
* `p2p/filter/id/<ID>`, where `<ID>` is the peer node ID (ie. the
  pubkey.Address() for the peer's PubKey)

> If either of these queries return a non-zero ABCI code, Tendermint will refuse

Author: I haven't touched the crash recovery section, as I am focusing on that in my next PR
